cpp/src/mlt/decode/int.cpp

```diff
@@ -0,0 +1,52 @@
+#include <mlt/decode/int.hpp>
+#include <mlt/decode/int_template.hpp>
+
+// from fastpfor/...
+#include <compositecodec.h>
+#include <fastpfor.h>
+#include <variablebyte.h>
+
+#include <cstdint>
+
+namespace mlt::decoder {
+
+struct IntegerDecoder::Impl {
+    FastPForLib::CompositeCodec<FastPForLib::FastPFor<8>, FastPForLib::VariableByte> codec;
+    // Reused storage for intermediate values
+    std::vector<std::uint32_t> buffer4;
+    std::vector<std::uint64_t> buffer8;
+};
+
+IntegerDecoder::IntegerDecoder() noexcept(false)
+    : impl(std::make_unique<Impl>()) {}
+
+IntegerDecoder::~IntegerDecoder() noexcept = default;
+
+template <typename T>
+std::vector<T>& IntegerDecoder::getTempBuffer() noexcept {
+    // Can we make these covariant and eliminate the casts?
+    if constexpr (std::is_trivially_assignable_v<decltype(impl->buffer4)::reference, T>) {
+        return reinterpret_cast<std::vector<T>&>(impl->buffer4);
+    } else if constexpr (std::is_trivially_assignable_v<decltype(impl->buffer8)::reference, T>) {
+        return static_cast<std::vector<T>&>(impl->buffer8);
+    }
+    // Ideally this would be a `static_assert`, but that fails even if never reached by template expansion.
```

Comment:
You could assign the if constexpr values to a constexpr bool and then do `static_assert(!a && !b)` right?

Reply:
Indeed it does, clever.  And at least clang 14 is smart enough to not complain about a missing return value.